src/Protocol.cpp

```diff
@@ -0,0 +1,156 @@
+#include <cstdlib>
+#include <sonar_oculus_m750d/Protocol.hpp>
+#include <string.h>
+
+using namespace sonar_oculus_m750d;
+
+void Protocol::handleBuffer(uint8_t const* buffer)
+{
+    OculusMessageHeader header;
+    memcpy(&header, buffer, sizeof(OculusMessageHeader));
+    switch (header.msgId) {
+        case messageSimplePingResult:
+            handleMessageSimplePingResult(buffer, header);
+            break;
+        case messagePingResult:
+            handleMessagePingResult(buffer);
+            break;
+        default:
+            break;
+    }
+}
+
+void setBearings(SonarData& sonar_data, uint32_t size, uint8_t const* buffer);
+void setImage(SonarData& sonar_data, uint32_t image_offset, uint8_t const* buffer);
+
+void Protocol::handleMessageSimplePingResult(uint8_t const* buffer,
+    OculusMessageHeader const& header)
+{
+    m_simple_ping_result = true;
+
+    uint32_t size = 0;
+    uint32_t image_offset = 0;
+
+    if (header.msgVersion == 2) {
+        OculusSimplePingResult2 result;
+        size = sizeof(OculusSimplePingResult2);
+        memcpy(&result, buffer, size);
+        m_data.image_size = result.imageSize;
+        image_offset = result.imageOffset;
+        m_data.beam_count = result.nBeams;
+        m_data.bin_count = result.nRanges;
+        m_data.range = m_data.bin_count * result.rangeResolution;
+        m_data.speed_of_sound = result.speedOfSoundUsed;
+    }
+    else {
+        OculusSimplePingResult result;
+        size = sizeof(OculusSimplePingResult);
+        memcpy(&result, buffer, size);
+        m_data.image_size = result.imageSize;
+        image_offset = result.imageOffset;
+        m_data.beam_count = result.nBeams;
+        m_data.bin_count = result.nRanges;
+        m_data.range = m_data.bin_count * result.rangeResolution;
+        m_data.speed_of_sound = result.speedOfSoundUsed;
+    }
+
+    setImage(m_data, image_offset, buffer);
+    setBearings(m_data, size, buffer);
+}
+
+void setImage(SonarData& sonar_data, uint32_t image_offset, uint8_t const* buffer)
+{
+    sonar_data.image = (uint8_t*)realloc(sonar_data.image, sonar_data.image_size);
+    if (sonar_data.image) {
+        memcpy(sonar_data.image, buffer + image_offset, sonar_data.image_size);
+    }
+}
+
+void setBearings(SonarData& sonar_data, uint32_t size, uint8_t const* buffer)
+{
+    sonar_data.bearings =
+        (short*)realloc(sonar_data.bearings, sonar_data.beam_count * sizeof(short));
+    if (sonar_data.bearings) {
+        memcpy(sonar_data.bearings, buffer + size, sonar_data.beam_count * sizeof(short));
+    }
+}
+
+void Protocol::handleMessagePingResult(uint8_t const* buffer)
```

Comment:
If we don't plan to neither have to handle this, please just remove the related code